### PR TITLE
rewrite and improve tests for truncated back-propagation

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Follow up to #9266 
Rewrites all tests for truncated backpropagation. There were only two tests but these were incomplete and not testing the core requirements. 
The major changes are: 
- Added Boring LSTM model, one for automatic and one for manual optimization
- parametrized tests with automatic and manual optimization
- split one big test into two smaller ones for unit testing the individual requirements: 1) verification of sequence splitting and passing outputs to the epoch_end hook  2) assertions for hidden state persistence
- use a an actual hidden state for the LSTM (not done previously)
- verify automatic detachment of hidden state
- moved the test file to the loops folder

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: tests/models/test_truncated_bptt.py [hunk 1]
```diff
@@ -74,104 +60,113 @@ def training_step(self, batch, batch_idx, hiddens):
...
+            self.test_hidden = None
 
-        def __len__(self):
-            return 1
+        def training_step(self, batch, batch_idx, hiddens):
+            split_idx = self.trainer.fit_loop.split_idx
+            # the hidden state may only be None for the first split_idx
+            assert split_idx == 0 and hiddens is None or split_idx != 0 and hiddens is not None
+            # test_hiddens is None when hiddens is None
+            assert not ((hiddens is None) ^ (self.test_hidden is None))
```
Comment (carmocca) [inline tests/models/test_truncated_bptt.py:81]: Could use XOR here too

and the parentheses aren't necessary

```suggestion
            assert not (split_idx == 0) ^ (hiddens is None)
            # test_hiddens is None when hiddens is None
            assert not (hiddens is None) ^ (self.test_hidden is None)
```

**Review by carmocca — APPROVED** [review]

File: tests/models/test_truncated_bptt.py (hunk 1 above)
Comment (awaelchli) [reply tests/models/test_truncated_bptt.py]: yes, added another xnor. 
I think parenthesis are ok here considering probably many people don't know `^` and precedence :)

**Review by SeanNaren — APPROVED** [review]
A very pretty set of tests @awaelchli :)

**Review by ananthsub — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
